File: sicp_1_19.swift
#!/usr/bin/env swift
// sicp exercise 1.19

func even(x: Int) -> Bool {
  return x % 2 == 0
}

func square(x: Int) -> Int {
  return x * x
}

func fib(n: Int) -> Int {
  return fibIter(1,0,0,1,n)
}

func fibIter(a: Int, b: Int, p: Int, q: Int, count: Int) -> Int {
  if count == 0 {
    return b
  } else if even(count) {
    return fibIter(a,b,(square(p)+square(q)),(square(q)+(2*p*q)),(count/2))
  } else {
    return fibIter((b*q)+(a*q)+(a*p),((b*p)+(a*q)),p,q,(count-1))
  }
}

for i in 1..<20 {
  print("\(fib(i)), ")
}
println()

File: sicp_1_19.js
#!/usr/bin/env node

function even(x) { return x % 2 === 0 }
function square(x) { return x * x }
function fib(x) { return fibIter(1,0,0,1,x) }

function fibIter(a,b,p,q,c) {
  if (c === 0) {
    return b;
  } else if (even(c)) {
    return fibIter(a,b,(square(p) + square(q)),(square(q) + (2*p*q)),(c/2|0));
  } else {
    return fibIter((b*q)+(a*q)+(a*p),((b*p)+(a*q)),p,q,(c-1));
  }
}

for (var i = 1; i <= 20; i++) {
  console.log(fib(i));
}